{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b2bfac0b-14b6-4811-887f-9fc6fae51203",
   "metadata": {},
   "source": [
    "### Overview\n",
    "\n",
    "This repo contains a script for coloring structural data from PDB based on RNA sequencing enrichment values. The basic workflow is as follows:\n",
    "\n",
    "    1. The sequences of the desired RNAs are pulled from the specific PDB structure as well as from the genomic FASTA file used for alignment.\n",
    "    2. The PDB and genomic sequences are aligned using Muscle. The alignment files are used to match nucleotide positions in the genomic file to those in the structure.\n",
    "    3. The new nucleotide positions are used to update a modified 1nt bedgraph containing the sequencing enrichment values and the identities of protein targets. Additionally, the bedgraph is binned and normalized according to user specifications. \n",
    "    4. A custom colormap is generated for each protein target and hex codes are generated for each enrichment value. A new file is generated containing the chain, residue, and color information that will style the PDB structure.\n",
    "    5. A ChimeraX readable script is generated based on a template and the colormapping information.\n",
    "    6. The user opens ChimeraX and executes the script. This will open the desired PDB structure, apply the styling specified in the template and colormapping files, and save a ChimeraX session file.\n",
    "\n",
    "All of these steps are wrapped in a single script: mapPdb.py. mapPdb.py takes a single argument: the seqInfo.yaml file. The yaml file supplies all necessary information for finding the correct PDB structure and applying colors as desired, and also points to the sequencing data.\n",
    "\n",
    "Sequencing data for mapping should be stored in a modified 1nt bedgraph file referred to here as a \"tidy bedgraph\". The tidy bedgraph has five columns:\n",
    "\n",
    "    - Column 1: RNA name \n",
    "    - Column 2: Start nucleotide of bin\n",
    "    - Column 3: End nucleotide of bin\n",
    "    - Column 4: Enrichment \n",
    "    - Column 5: Experimental target\n",
    "\n",
    "The tidy bedgraph format allows for multiple targets to be mapped to the same structure. For example, if you have CLIP data for RPS5 and RPL36, two tidy bedgraphs can be concatenated without losing the information of which enrichment values belong to each protein.\n",
    "\n",
    "mapPdb is a relatively simple tool for comparing sequencing and structural data. The ChimeraX script is usually generated in a few seconds, making it easy to adjust styling preferences in the seqInfo.yaml or chimeraxTemplate.py files and rerun the pipeline. From there, publication quality figures can be generated using standard ChimeraX tools.\n",
    "\n",
    "### Getting started\n",
    "\n",
    "The final ChimeraX script generated is compatible with ChimeraX 1.8 but has not been tested on other versions. ChimeraX can be downloaded [here](https://www.cgl.ucsf.edu/chimerax/download.html). \n",
    "\n",
    "After cloning this repo, create the mapPdb Conda environment. I recommend using [Mamba](https://mamba.readthedocs.io/en/latest/installation/mamba-installation.html) for package management. On most systems, run:\n",
    "\n",
    "```mamba env create -f env/mapPdb.yaml```\n",
    "\n",
    "Then, before running the script, run:\n",
    "\n",
    "```conda activate mapPdb```\n",
    "\n",
    "### Running the example\n",
    "\n",
    "To make sure that the installation is working, run the example pipeline using:\n",
    "\n",
    "```python mapPdb.py example/4ug0.yaml```\n",
    "\n",
    "This should output a folder called `4ug0/`. In `4ug0/` is a script called `4ug0_chimeraX.py`. Open ChimeraX and run the script from the ChimeraX command line:\n",
    "\n",
    "`run /absolute/path/to/mapPdb/4ug0/4ug0_chimeraX.py`\n",
    "\n",
    "ChimeraX requires absolute paths for executing scripts. If everything has run correctly, the 4UG0 structure should appear with protein and RNA coloration as specified in the `4ug0.yaml` config file. In the `4ug0/` directory the ChimeraX session is now stored as `4ug0_sesh.cxs`.\n",
    "\n",
    "Note that all chains other than those specified in the config file have been deleted. In contrast, all RNA residues that did not receive a color have been rendered invisible. To reveal the surfaces of those bases, run:\n",
    "\n",
    "`show /S2 /L5 surfaces`\n",
    "\n",
    "in the ChimeraX command line. The surfaces can be hidden using:\n",
    "\n",
    "`hide /S2 /L5 surfaces`\n",
    "\n",
    "For a guide on other commands for styling and saving images, consult the [ChimeraX documentation](https://www.cgl.ucsf.edu/chimerax/docs/user/index.html).\n",
    "\n",
    "### Customizing the config file\n",
    "\n",
    "The file `seqInfo.yaml` contains instructions on how to adjust the config file for a new visualization. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30a1c86d-afde-4c74-b42f-74bbe13d592d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
